Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+HTTP1State.swift

```diff
@@ -0,0 +1,694 @@
+                }
+
+                // the cleanupAction here is pretty lazy :)
+                return .init(.none, .cleanupConnection(close: [], cancel: [], isShutdown: isShutdown))
+
+            case .shutDown:
+                preconditionFailure("The pool is already shutdown all connections must already been torn down")
+            }
+        }
+
+        mutating func connectionTimeout(_ connectionID: Connection.ID) -> Action {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                // because of a race this connection (connection close runs against trigger of timeout)
+                // was already removed from the state machine.
+                return .init(.none, .none)
+            }
+
+            assert(self.state == .running, "If we are shutting down, we must not have any idle connections")
+
+            var connectionState = self.connections[index]
+            guard connectionState.isAvailable else {
+                // connection is not available anymore, we may have just leased it for a request
+                return .init(.none, .none)
+            }
+
+            assert(self.waiters.isEmpty, "We have an idle connection, that times out, but waiters? Something is very wrong!")
+
+            self.connections.remove(at: index)
+            return .init(.none, .closeConnection(connectionState.close(), isShutdown: .no))
+        }
+
+        mutating func http1ConnectionReleased(_ connectionID: Connection.ID) -> Action {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("A connection that we don't know was released? Something is very wrong...")
+            }
+
+            var connectionState = self.connections[index]
+            connectionState.release()
+
+            switch self.state {
+            case .running:
+                guard let nextWaiter = self.waiters.popFirst() else {
+                    // there is no more work todo immidiatly
+                    self.connections[index] = connectionState
+                    return .init(.none, .scheduleTimeoutTimer(connectionID))
+                }
+
+                assert(self.connections.count == self.maximumConcurrentConnections,
+                       "Why do we have waiters, if we could open more connections?")
+
+                guard nextWaiter.canBeRun(on: connectionState.eventLoop) else {
+                    let eventLoop = nextWaiter.eventLoopRequirement!
+                    let newConnection = HTTP1ConnectionState(
+                        connectionID: self.idGenerator.next(),
+                        eventLoop: eventLoop,
+                        waiter: nextWaiter
+                    )
+                    self.connections[index] = newConnection
+                    return .init(.none, .replaceConnection(connectionState.close(), with: newConnection.connectionID, on: eventLoop))
```

Comment:
This logic appears to be repeated in a number of places. Consider factoring it out.